### PR TITLE
Binary Search Tree

### PR DESCRIPTION
**Pull request by DuckNrOne**
Hello,
I implemented the A Binary Search Tree.
I annouced it in #112.

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have made corresponding changes to the comments
- [x] I have made corresponding changes to the README.md

### CODE REVIEW

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- DataStructures/BinarySearchTree/Node.cs  1
- DataStructures.Tests/BinarySearchTreeTests.cs  1
- DataStructures/BinarySearchTree/BinarySearchTree.cs  6
         

Clones added
============
- DataStructures/BinarySearchTree/BinarySearchTree.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6278013&bid=20433197)

File: DataStructures.Tests/BinarySearchTreeTests.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            Assert.AreEqual(expectedList, testTree.Inorder());
+            expectedList = new List<int>(new int [] {3, 4, 1, 7, 5, 9, 8});
+            Assert.AreEqual(expectedList, testTree.Postorder());   
+        }
+
+        [Test, Order(2)]
+        public void Insert_Inorder_Test()
+        {
+            testTree.Insert(4);
+            List<int> expectedList = new List<int>(new int [] {1, 3, 4, 4, 5, 7, 8, 9});
```
Comment (siriak) [inline DataStructures.Tests/BinarySearchTreeTests.cs:40]: Codacy found an issue: [Remove the array type; it is redundant.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6278013)

File: DataStructures/BinarySearchTree/BinarySearchTree.cs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+            else
+            {
+                // If it isn't smaller or bigger it equals the current node
+
+                // Node with only one child
+                if (node.Left is null)
+                {
+                    return node.Right;
+                }
+                else if (node.Right is null)
```
Comment (siriak) [inline DataStructures/BinarySearchTree/BinarySearchTree.cs:172]: Codacy found an issue: [Add the missing 'else' clause.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6278013)

File: DataStructures/BinarySearchTree/BinarySearchTree.cs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    {
+        /// <summary>
+        /// Current root of a subtree.
+        /// </summary>
+        private Node<T>? rootNode;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BinarySearchTree{T}"/> class.
+        /// </summary>
+        public BinarySearchTree()
```
Comment (siriak) [inline DataStructures/BinarySearchTree/BinarySearchTree.cs:25]: Codacy found an issue: [Remove this redundant constructor.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6278013)

File: DataStructures.Tests/BinarySearchTreeTests.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            testTree.Insert(9);
+            testTree.Insert(3);
+        }
+
+        [Test, Order(1)]
+        public void Order_Test()
+        {
+            List<int> expectedList = new List<int>(new int [] {8, 5, 1, 4, 3, 7, 9});
+            Assert.AreEqual(expectedList, testTree.Preorder());
+            expectedList = new List<int>(new int [] {1, 3, 4, 5, 7, 8, 9});
```
Comment (siriak) [inline DataStructures.Tests/BinarySearchTreeTests.cs:30]: Codacy found an issue: [Remove the array type; it is redundant.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6278013)

File: DataStructures.Tests/BinarySearchTreeTests.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            Assert.AreEqual(expectedList, testTree.Inorder());
+            expectedList = new List<int>(new int [] {3, 4, 1, 7, 5, 9, 8});
```
Comment (siriak) [inline DataStructures.Tests/BinarySearchTreeTests.cs:32]: Codacy found an issue: [Remove the array type; it is redundant.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6278013)

File: DataStructures.Tests/BinarySearchTreeTests.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            testTree.Insert(4);
+            testTree.Insert(7);
+            testTree.Insert(9);
+            testTree.Insert(3);
+        }
+
+        [Test, Order(1)]
+        public void Order_Test()
+        {
+            List<int> expectedList = new List<int>(new int [] {8, 5, 1, 4, 3, 7, 9});
```
Comment (siriak) [inline DataStructures.Tests/BinarySearchTreeTests.cs:28]: Codacy found an issue: [Remove the array type; it is redundant.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6278013)

File: DataStructures.Tests/BinarySearchTreeTests.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            testTree.Insert(4);
+            List<int> expectedList = new List<int>(new int [] {1, 3, 4, 4, 5, 7, 8, 9});
+            Assert.AreEqual(expectedList, testTree.Inorder());
+        }
+
+        [Test, Order(3)]
+        public void Remove_Postorder_Test()
+        {
+            testTree.Remove(3);
+            List<int> expectedList = new List<int>(new int [] {4, 4, 1, 7, 5, 9, 8});
```
Comment (siriak) [inline DataStructures.Tests/BinarySearchTreeTests.cs:48]: Codacy found an issue: [Remove the array type; it is redundant.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6278013)

File: DataStructures.Tests/BinarySearchTreeTests.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+using DataStructures.BinarySearchTree;
+using NUnit.Framework;
+using System.Collections.Generic;
+
+namespace DataStructures.Tests
+{
+    [TestFixture]
+    class BinarySearchTreeTests
+    {
+        BinarySearchTree<int> testTree = new BinarySearchTree<int>();
```
Comment (siriak) [inline DataStructures.Tests/BinarySearchTreeTests.cs:11]: Codacy found an issue: [Make 'testTree' 'readonly'.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6278013)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures/BinarySearchTree/Node.cs  1
- DataStructures.Tests/BinarySearchTreeTests.cs  1
- DataStructures/BinarySearchTree/BinarySearchTree.cs  6
         

Clones added
============
- DataStructures/BinarySearchTree/BinarySearchTree.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6278013&bid=20433197)

**Review by siriak — COMMENTED** [review]
Hi, there are some merge conflicts, could you resolve them?

**Comment by DuckNrOne** [comment]
Oh :D I didn't see that there is already an implementation for biary search tree. upsi :D

**Comment by siriak** [comment]
Closing this one, feel free to open another one when needed

**Pull request closed without merging**